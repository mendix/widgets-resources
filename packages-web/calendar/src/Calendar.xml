<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.calendar.Calendar" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/">
    <name>Calendar</name>
    <description>Display and manage calendar events</description>
    <icon/>
    <properties>
        <property key="defaultView" type="enumeration" defaultValue="month">
            <caption>Default view</caption>
            <category>Appearance</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="month">Month</enumerationValue>
                <enumerationValue key="week">Week</enumerationValue>
                <enumerationValue key="work_week">Work week</enumerationValue>
                <enumerationValue key="day">Day</enumerationValue>
                <enumerationValue key="agenda">Agenda</enumerationValue>
            </enumerationValues>
        </property>
        <property key="popup" type="boolean" defaultValue="true">
            <caption>Popup</caption>
            <category>Appearance</category>
            <description>Show truncated events in an overlay when you click the "+x more" link</description>
        </property>
        <property key="selectable" type="boolean" defaultValue="true">
            <caption>Selectable</caption>
            <category>Appearance</category>
            <description/>
        </property>
        <property key="editable" type="enumeration" defaultValue="default">
            <caption>Editable</caption>
            <category>Appearance</category>
            <description>The default option is based on the page settings and the attribute's write access</description>
            <enumerationValues>
                <enumerationValue key="default">Default</enumerationValue>
                <enumerationValue key="never">Never</enumerationValue>
            </enumerationValues>
        </property>
        <property key="dataSource" type="enumeration" defaultValue="XPath">
            <caption>Data source</caption>
            <category>Event data</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="XPath">Database</enumerationValue>
                <enumerationValue key="microflow">Microflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="eventEntity" type="entity" required="false">
            <caption>Event entity</caption>
            <category>Event data</category>
            <description/>
        </property>
        <property key="entityConstraint" type="entityConstraint" entityProperty="eventEntity" required="false">
            <caption>XPath</caption>
            <category>Event data</category>
            <description/>
        </property>
		<property key="dataSourceMicroflow" type="microflow" required="false">
			<caption>Data source microflow</caption>
			<category>Event data</category>
			<description></description>
			<returnType type="Void" />
		</property>
        <property key="titleAttribute" type="attribute" entityProperty="eventEntity" isPath="optional" pathType="reference" required="false">
            <caption>Title</caption>
            <category>Event data</category>
            <description/>
            <attributeTypes>
                <attributeType name="String"/>
            </attributeTypes>
        </property>
        <property key="startAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>Start</caption>
            <category>Event data</category>
            <description/>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="endAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>End</caption>
            <category>Event data</category>
            <description/>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="onClickEvent" type="enumeration" defaultValue="doNothing">
            <caption>On click event</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="eventPage" type="form" required="false">
            <caption>Page</caption>
            <category>Events</category>
            <description>The page that will be opened on clicking an event</description>
        </property>
        <property key="openPageAs" type="enumeration" defaultValue="content">
            <caption>Open page as</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="content">Full page</enumerationValue>
                <enumerationValue key="popup">Popup</enumerationValue>
                <enumerationValue key="modal">Blocking popup</enumerationValue>
            </enumerationValues>
        </property>
        <property key="eventMicroflow" type="microflow" required="false">
            <caption>Microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on clicking an event</description>
            <returnType type="Void" />
        </property>
        <property key="eventNanoflow" type="nanoflow" required="false">
            <caption>Nanoflow</caption>
            <category>Events</category>
            <description>The nanoflow that will be executed on clicking an event</description>
            <returnType type="Void" />
        </property>
        <property key="onClickSlotEvent" type="enumeration" defaultValue="doNothing">
            <caption>On click slot</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="slotPage" type="form" required="false">
            <caption>New event page</caption>
            <category>Events</category>
            <description>The page that will be opened on clicking a slot</description>
        </property>
        <property key="slotMicroflow" type="microflow" required="false">
            <caption>New event microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on clicking a slot</description>
            <returnType type="Void" />
        </property>
        <property key="slotNanoflow" type="nanoflow" required="false">
            <caption>New event nanoflow</caption>
            <category>Events</category>
            <description>The nanoflow that will be executed on clicking a slot</description>
            <returnType type="Void" />
        </property>
        <property key="onDropEvent" type="enumeration" defaultValue="doNothing">
            <caption>On drop event</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="onDropPage" type="form" required="false">
            <caption>On drop page</caption>
            <category>Events</category>
            <description>The page that will be opened on droping an event</description>
        </property>
        <property key="onDropMicroflow" type="microflow" required="false">
            <caption>On drop microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on droping an event</description>
            <returnType type="Void" />
        </property>
        <property key="onDropNanoflow" type="nanoflow" required="false">
            <caption>On drop nanoflow</caption>
            <category>Events</category>
            <description>The nanoflow that will be executed on droping an event</description>
            <returnType type="Void" />
        </property>
    </properties>
</widget>
