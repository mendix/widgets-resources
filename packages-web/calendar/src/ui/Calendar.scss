.widget-calendar {
    .rbc-calendar {
        min-height: 580px !important;
        height: 580px;
    }
}
