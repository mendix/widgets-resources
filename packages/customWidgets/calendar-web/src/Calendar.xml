<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.calendar.Calendar" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Calendar</name>
    <description>Display and manage calendar events</description>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAAAXNSR0IArs4c6QAAA69JREFUeAHtWs1uEzEQtjeRKBISBw5wKD9vAVc4QFcgVA4hZyRoX6OvQUHiHOVABUKBC1zpYyCCgEMOIKQGKV0zs1qn3l27tnftrUO8UhV7djw/n2fGP11C4hMRiAisMwK0qfPD4TDNsmwfxm8qZEyTJNkZjUYTxXsjsm89iZEVEiaN8zhis+CRjDYn+dbTGAB00MANEx6dGBMZJjxSPW0AkApcNWKtBhjknNTH8XicyxoMBkzK4IjYQo+0JtUiwCDnHLnSuRhpTaoBAGY1zqfOXbJXWPNNBoC92BUe0dfZznOuymea66rxVXmqfls9uvExAlTI29J1SNvKU/G71rP2EdAGgKlqlgS6CY/ALm2ayDDhkQpvDAAedEDiaYrzjYdUqwXRt56+hS0l1uKUd7VE9NDxradxBHjw9UxERgDOBPaAlGprgOt1NyDfc1NiCoQ2I13bs/YRoK0BbU9zXc9oVZ+uhq19BEQAqiGzbn1tDWgLyN4n1j/88/cOOyYDSllKGFkkfbr77sHGB5SdvpmnJGP7cJVcu6+T6Yar5y/ieBmPDc0LAKLTn3/NH4FBl9AoVlyYZwv2HLo3kGbjPLKDiOul8Uhs8TgHAGf08Pf8FTh7WWUXOsHfQdto5jl/8dur9Bt33RdBxl6c5nxjS4WBjND3QrdV03kEgPOZxKIZ0PI0qL6DnJ7yKMD2ZPt86Y4B0wnS6Ic4nvbIuCqnad95BNCE7lJKvoFB3wmhL2mP3rt1ceOK0sCE7qDj+EegXeXDAgo0EbzZzQvnPlb5mvadR8Dk4QZ+D1DKayRsHRxJbSz4S7MuMtKMPIYIER76eu82XQiEVk3nEdDKmspgDP+MsW2R7DL8UW7QAPgO/+ABwM2TOPtQU5yGP8p2XgPKBpv35DvCcvbDNuhpenC0hcWyqB3mChSc4aSA4XY4XzKBV+GPNTkcAKxNdzMgHACK/YDOLdV+QTdO9T6YGiDuB9K387vsmOWnxcLwGW6mXK7/HJBwIoBbBL9dVH+uLjgAcPMDEOARevm43vwsBUMjOAC62PyIAHirAfJ1/UQ1nA3yRZ4XNb6udxn+aI2/CGi8rrP7JzDB3s/h0VeUy9v+AOAaLH/hKL28T4D2T5dHX5kp/gBouq5T+gwM/ZrfKVD6xMfSJwLhrQaI67qoUNcuxl3T8bl67y8CXFnoWU4EwDPAwYvX1gDdf1eD91BjYEwBCUBTCe1/IdV8q0WAwZeZqwqGky9XV9X5aHdEICIgR+AfGqgwglFLPQMAAAAASUVORK5CYII=</icon>
    <properties>
        <property key="dataSource" type="enumeration" defaultValue="context">
            <caption>Data source</caption>
            <category>Data source</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="context">Context</enumerationValue>
                <enumerationValue key="XPath">Database</enumerationValue>
                <enumerationValue key="microflow">Microflow</enumerationValue>
                <enumerationValue key="nanoflow">Nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="eventEntity" type="entity" allowNonPersistableEntities="true">
            <caption>Event entity</caption>
            <category>Data source</category>
            <description>With data context, this entity should be equal to the containing data view</description>
        </property>
        <property key="entityConstraint" type="entityConstraint" entityProperty="eventEntity" required="false">
            <caption>XPath</caption>
            <category>Data source</category>
            <description>Also allows the event objects to be constrained to a relation to context by using [%CurrentObject%]</description>
        </property>
        <property key="dataSourceMicroflow" type="microflow" required="false">
            <caption>Microflow</caption>
            <category>Data source</category>
            <description/>
            <returnType type="Object" isList="true" entityProperty="eventEntity"/>
        </property>
        <property key="dataSourceNanoflow" type="nanoflow" required="false">
            <caption>Nanoflow</caption>
            <category>Data source</category>
            <description/>
            <returnType type="Object" isList="true" entityProperty="eventEntity"/>
        </property>
        <property key="titleAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>Title attribute</caption>
            <category>Data source</category>
            <description/>
            <attributeTypes>
                <attributeType name="String"/>
            </attributeTypes>
        </property>
        <property key="allDayAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>All day attribute</caption>
            <category>Data source</category>
            <description/>
            <attributeTypes>
                <attributeType name="Boolean"/>
            </attributeTypes>
        </property>
        <property key="startAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>Start attribute</caption>
            <category>Data source</category>
            <description/>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="endAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>End attribute</caption>
            <category>Data source</category>
            <description/>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="eventColor" type="attribute" entityProperty="eventEntity" required="false">
            <caption>Color attribute</caption>
            <category>Data source</category>
            <description>Attribute containing a valid html color eg: red #FF0000 rgb(250,10,20) rgba(10,10,10, 0.5)</description>
            <attributeTypes>
                <attributeType name="Enum" />
                <attributeType name="String" />
            </attributeTypes>
        </property>
        <property key="executeOnViewChange" type="boolean" defaultValue="false">
            <caption>Refresh data source on view change</caption>
            <category>Data source</category>
            <description>For data source microflow, enable to retrieve a sub set of large data set based on view start and end attribute</description>
        </property>
        <property key="viewStartAttribute" type="attribute" required="false" allowNonPersistableEntities="true">
            <caption>View start attribute</caption>
            <category>Data source</category>
            <description>The date on which the calendar view starts, will be updated with `Refresh data source` before the view changes</description>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="viewEndAttribute" type="attribute" required="false" allowNonPersistableEntities="true">
            <caption>View end attribute</caption>
            <category>Data source</category>
            <description>The date on which the calendar view ends, will be updated with `Refresh data source` before the view changes</description>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="view" type="enumeration" defaultValue="standard">
            <caption>View</caption>
            <category>View</category>
            <description>Standard has Day, Week and Month</description>
            <enumerationValues>
                <enumerationValue key="standard">Standard</enumerationValue>
                <enumerationValue key="custom">Custom</enumerationValue>
            </enumerationValues>
        </property>
        <property key="defaultView" type="enumeration" defaultValue="month">
            <caption>Initial selected view</caption>
            <category>View</category>
            <description>Work week and agenda are only available in custom views</description>
            <enumerationValues>
                <enumerationValue key="day">Day</enumerationValue>
                <enumerationValue key="week">Week</enumerationValue>
                <enumerationValue key="month">Month</enumerationValue>
                <enumerationValue key="work_week">(Work week)</enumerationValue>
                <enumerationValue key="agenda">(Agenda)</enumerationValue>
            </enumerationValues>
        </property>
        <property key="startDateAttribute" type="attribute" isPath="optional" pathType="reference" required="false">
            <caption>Start date attribute</caption>
            <category>View</category>
            <description>The start date that should be shown in the view</description>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="customViews" type="object" isList="true" required="false">
            <caption>Custom top bar views</caption>
            <category>View</category>
            <description/>
            <properties>
                <property key="customView" type="enumeration" defaultValue="month">
                    <caption>Item</caption>
                    <category>Appearance</category>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="previous">Previous button</enumerationValue>
                        <enumerationValue key="today">Today button</enumerationValue>
                        <enumerationValue key="next">Next button</enumerationValue>
                        <enumerationValue key="title">Title date text</enumerationValue>
                        <enumerationValue key="month">Month button</enumerationValue>
                        <enumerationValue key="week">Week button</enumerationValue>
                        <enumerationValue key="work_week">Work week button</enumerationValue>
                        <enumerationValue key="day">Day button</enumerationValue>
                        <enumerationValue key="agenda">Agenda button</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="position" type="enumeration" defaultValue="left">
                    <caption>Position</caption>
                    <category>Appearance</category>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="left">Left</enumerationValue>
                        <enumerationValue key="center">Center</enumerationValue>
                        <enumerationValue key="right">Right</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="customCaption" type="translatableString" required="false">
                    <caption>Caption</caption>
                    <category>Appearance</category>
                    <description>The text for the button or title. For type 'Title date text' this field will be replaced by defaults according to the language and view.</description>
                </property>
                <property key="renderMode" type="enumeration" defaultValue="button">
                    <caption>Render mode</caption>
                    <category>Appearance</category>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="button">Button</enumerationValue>
                        <enumerationValue key="link">Link</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="buttonToolTip" type="translatableString" required="false">
                    <caption>Button tooltip</caption>
                    <category>Appearance</category>
                    <description/>
                </property>
                <property key="buttonStyle" type="enumeration" defaultValue="default">
                    <caption>Button style</caption>
                    <category>Appearance</category>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="default">Default</enumerationValue>
                        <enumerationValue key="primary">Primary</enumerationValue>
                        <enumerationValue key="success">Success</enumerationValue>
                        <enumerationValue key="info">Info</enumerationValue>
                        <enumerationValue key="inverse">Inverse</enumerationValue>
                        <enumerationValue key="warning">Warning</enumerationValue>
                        <enumerationValue key="danger">Danger</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="headerFormat" type="translatableString" required="false">
                    <caption>Header day format</caption>
                    <category>Custom formats</category>
                    <description>Format of date(s) in the header above the day, week, month, work week or Agenda view. Use Mendix date formats. e.g. 'EEEE dd/MM'</description>
                </property>
                <property key="cellDateFormat" type="translatableString" required="false">
                    <caption>Cell date format</caption>
                    <category>Custom formats</category>
                    <description>Date shown in the month cells. Use Mendix date formats. e.g. 'dd'</description>
                    <translations>
                        <translation lang="en_US">dd</translation>
                        <translation lang="nl_NL">dd</translation>
                    </translations>
                </property>
                <property key="gutterTimeFormat" type="translatableString" required="false">
                    <caption>Time gutter format</caption>
                    <category>Custom formats</category>
                    <description>Time shown as the first column in the week, day and agenda view. Use Mendix date formats. e.g. HH:mm</description>
                    <translations>
                        <translation lang="en_US">hh:mm a</translation>
                        <translation lang="nl_NL">HH:mm</translation>
                    </translations>
                </property>
                <property key="gutterDateFormat" type="translatableString" required="false">
                    <caption>Date gutter format</caption>
                    <category>Custom formats</category>
                    <description>Date shown as the first column in the agenda view. e.g. 'EEE MMM d'</description>
                    <translations>
                        <translation lang="en_US">EEE dd/MM</translation>
                        <translation lang="nl_NL">EEE MM-dd</translation>
                    </translations>
                </property>
                <property key="allDayText" type="translatableString" required="false">
                    <caption>Default all day text</caption>
                    <category>Text</category>
                    <description>Text showing in the agenda view to indicate an event will be all day</description>
                    <translations>
                        <translation lang="en_US">All day</translation>
                        <translation lang="nl_NL">Gehele dag</translation>
                    </translations>
                </property>
                <property key="textHeaderDate" type="translatableString" required="false">
                    <caption>Header date</caption>
                    <category>Text</category>
                    <description>Text showing in the agenda view to header in the column date</description>
                    <translations>
                        <translation lang="en_US">Date</translation>
                        <translation lang="nl_NL">Datum</translation>
                    </translations>
                </property>
                <property key="textHeaderTime" type="translatableString" required="false">
                    <caption>Header time</caption>
                    <category>Text</category>
                    <description>Text showing in the agenda view to header in the column time</description>
                    <translations>
                        <translation lang="en_US">Time</translation>
                        <translation lang="nl_NL">Tijd</translation>
                    </translations>
                </property>
                <property key="textHeaderEvent" type="translatableString" required="false">
                    <caption>Header event</caption>
                    <category>Text</category>
                    <description>Text showing in the agenda view to header in the column event</description>
                    <translations>
                        <translation lang="en_US">Event</translation>
                        <translation lang="nl_NL">Afspraak</translation>
                    </translations>
                </property>
            </properties>
        </property>
        <property key="popup" type="boolean" defaultValue="true">
            <caption>Show truncated events</caption>
            <category>View</category>
            <description>Show truncated events in an overlay when you click the "+x more" link</description>
        </property>
        <property key="editable" type="enumeration" defaultValue="default">
            <caption>Editable</caption>
            <category>View</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="default">Default</enumerationValue>
                <enumerationValue key="never">Never</enumerationValue>
            </enumerationValues>
        </property>
        <property key="enableCreate" type="boolean" defaultValue="true">
            <caption>Enable create</caption>
            <category>View</category>
            <description/>
        </property>
        <property key="onClickEvent" type="enumeration" defaultValue="doNothing">
            <caption>On click</caption>
            <category>Events</category>
            <description>The click event is triggered when an exiting item is clicked on the calendar</description>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="onClickMicroflow" type="microflow" required="false" entityProperty="eventEntity">
            <caption>On click microflow</caption>
            <category>Events</category>
            <description/>
            <returnType type="Void" />
        </property>
        <property key="onClickNanoflow" type="nanoflow" required="false" entityProperty="eventEntity">
            <caption>On click nanoflow</caption>
            <category>Events</category>
            <description/>
            <returnType type="Void" />
        </property>
        <property key="onCreate" type="enumeration" defaultValue="doNothing">
            <caption>On create</caption>
            <category>Events</category>
            <description>The create event is triggered when a time slot is selected, and the 'Enable create' property is set to 'true'</description>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="newEventContextPath" type="entity" isPath="yes" allowNonPersistableEntities="true" pathType="reference" entityProperty="eventEntity" required="false">
            <caption>Create association</caption>
            <category>Events</category>
            <description>Association is set on create with the context</description>
        </property>
        <property key="onCreateMicroflow" type="microflow" required="false" entityProperty="eventEntity">
            <caption>On create microflow</caption>
            <category>Events</category>
            <description/>
            <returnType type="Void" />
        </property>
        <property key="onCreateNanoflow" type="nanoflow" required="false" entityProperty="eventEntity">
            <caption>On create nanoflow</caption>
            <category>Events</category>
            <description/>
            <returnType type="Void" />
        </property>
        <property key="onChangeEvent" type="enumeration" defaultValue="doNothing">
            <caption>On change</caption>
            <category>Events</category>
            <description>The change event is triggered on moving/dragging an item or changing the start or end time of by resizing an item</description>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="onChangeMicroflow" type="microflow" required="false" entityProperty="eventEntity">
            <caption>On change microflow</caption>
            <category>Events</category>
            <description/>
            <returnType type="Void" />
        </property>
        <property key="onChangeNanoflow" type="nanoflow" required="false" entityProperty="eventEntity">
            <caption>On change nanoflow</caption>
            <category>Events</category>
            <description/>
            <returnType type="Void" />
        </property>
        <property key="widthUnit" type="enumeration" defaultValue="percentage">
            <caption>Width unit</caption>
            <category>Size</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="percentage">Percentage</enumerationValue>
                <enumerationValue key="pixels">Pixels</enumerationValue>
            </enumerationValues>
        </property>
        <property key="width" type="integer" defaultValue="100">
            <caption>Width</caption>
            <category>Size</category>
            <description/>
        </property>
        <property key="heightUnit" type="enumeration" defaultValue="pixels">
            <caption>Height unit</caption>
            <category>Size</category>
            <description>'Percentage of width' is the aspect ratio, 'Pixels' is absolute. Warning: When using 'Percentage of parent' the parent container must have an absolute height, else nothing is displayed.</description>
            <enumerationValues>
                <enumerationValue key="percentageOfWidth">Percentage of width</enumerationValue>
                <enumerationValue key="pixels">Pixels</enumerationValue>
                <enumerationValue key="percentageOfParent">Percentage of parent</enumerationValue>
            </enumerationValues>
        </property>
        <property key="height" type="integer" defaultValue="580">
            <caption>Height</caption>
            <category>Size</category>
            <description/>
        </property>
    </properties>
</widget>
