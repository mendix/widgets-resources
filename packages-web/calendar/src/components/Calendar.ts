import { CSSProperties, Component, ReactChild, createElement, ReactNode } from "react";

import { Alert } from "./Alert";
import { Container, Style } from "../utils/namespaces";
import classNames from "classnames";
import BigCalendar from "react-big-calendar";
import moment from "moment";
import withDragAndDrop from "react-big-calendar/lib/addons/dragAndDrop";
import CustomToolbar from "./Toolbar";
import { SizeContainer } from "./SizeContainer";

import { CalendarLoader } from "./CalendarLoader";

import "react-big-calendar/lib/addons/dragAndDrop/styles.css";
import "react-big-calendar/lib/css/react-big-calendar.css";
import "../ui/Calendar.scss";
import "../ui/CalendarLoader.scss";

const localizer = BigCalendar.momentLocalizer(moment);
const DragAndDropCalendar = withDragAndDrop(BigCalendar);

export interface CalendarProps {
    alertMessage?: ReactChild;
    className?: string;
    events: CalendarEvent[];
    color?: string;
    formats?: {};
    enableCreate: boolean;
    height: number;
    heightUnit: Style.HeightUnitType;
    defaultView: Style.View;
    loading?: boolean;
    startPosition?: Date;
    messages: {};
    popup: boolean;
    editable: string;
    titleFormat?: (date: Date) => void;
    weekdayFormat?: (date: Date) => void;
    timeGutterFormat?: (date: Date) => void;
    monthHeaderFormat?: (date: Date) => void;
    dayHeaderFormat?: (date: Date) => void;
    style: CSSProperties;
    viewOption: "custom" | "standard";
    width: number;
    widthUnit: Style.WidthUnitType;
    onRangeChangeAction?: (date: object) => void;
    onSelectEventAction?: (eventInfo: object) => void;
    onEventResizeAction?: (eventInfo: object) => void;
    onSelectSlotAction?: (slotInfo: object) => void;
    onEventDropAction?: (eventInfo: object) => void;
    customViews: Container.CustomViews[];
}

interface HOCToolbarProps {
    customViews: Container.CustomViews[] | Array<Partial<Container.ButtonConfig>>;
    onClickToolbarButton: (date: object) => void;
}

export interface CalendarEvent {
    allDay: boolean;
    color: string;
    end: Date;
    guid: string;
    start: Date;
    title: string;
}

interface State {
    date?: Date;
}

class Calendar extends Component<CalendarProps, State> {
    readonly state: State = {
        date: undefined
    };

    private readonly onNavigateHandler = this.onNavigate.bind(this);

    render(): ReactNode {
        return createElement(
            SizeContainer,
            {
                className: classNames("widget-calendar", this.props.className),
                style: this.props.style,
                widthUnit: this.props.widthUnit,
                width: this.props.width,
                heightUnit: this.props.heightUnit,
                height: this.props.height
            },
            this.renderAlert(),
            this.renderCalendar()
        );
    }

    private getDefaultToolbar(): Array<Partial<Container.ButtonConfig>> {
        return [
            {
                customView: "previous",
                position: "left",
                buttonToolTip: "previous"
            },
            {
                customView: "today",
                position: "left",
                customCaption: "today"
            },
            {
                customView: "next",
                position: "left",
                buttonToolTip: "next"
            },
            {
                customView: "title",
                position: "center"
            },
            {
                customView: "day",
                position: "right",
                customCaption: "day"
            },
            {
                customView: "week",
                position: "right",
                customCaption: "week"
            },
            {
                customView: "month",
                position: "right",
                customCaption: "month"
            }
        ];
    }

    private renderAlert(): ReactNode {
        return createElement(Alert, { className: "widget-calendar-alert" }, this.props.alertMessage);
    }

    private onNavigate(date: Date): void {
        this.setState({ date });
    }

    private renderCalendar(): ReactNode {
        const wrapToolbar = (injectedProps: HOCToolbarProps): Function => (toolbarProps: Container.ToolbarProps) =>
            createElement(CustomToolbar as any, { ...injectedProps, ...toolbarProps });

        const props = {
            localizer,
            events: this.props.events,
            allDayAccessor: this.allDayAccessor,
            components: {
                toolbar: wrapToolbar({ customViews: this.getToolbarProps(), onClickToolbarButton: this.onRangeChange })
            },
            onNavigate: this.onNavigateHandler,
            eventPropGetter: this.eventColor,
            date: this.state.date ?? this.props.startPosition,
            defaultView: this.defaultView(),
            formats: this.props.viewOption === "custom" ? this.props.formats : "",
            messages: this.props.viewOption === "custom" ? this.props.messages : "",
            popup: this.props.popup,
            selectable: this.props.enableCreate,
            step: 60,
            showMultiDayTimes: true,
            onRangeChange: this.onRangeChange,
            onSelectEvent: this.onSelectEvent,
            onSelectSlot: this.onSelectSlot,
            views: ["month", "day", "week", "work_week", "month", "agenda"]
        };

        if (this.props.loading) {
            return createElement(CalendarLoader);
        } else if (this.props.enableCreate && this.props.editable === "default") {
            return createElement(DragAndDropCalendar, {
                ...props,
                onEventDrop: this.onEventDrop,
                onEventResize: this.onEventResize
            });
        } else {
            return createElement(BigCalendar, props);
        }
    }

    private defaultView(): Style.View {
        if (
            this.props.viewOption === "standard" &&
            (this.props.defaultView === "work_week" || this.props.defaultView === "agenda")
        ) {
            return "month";
        }

        return this.props.defaultView;
    }

    private getToolbarProps(): Array<Partial<Container.ButtonConfig>> {
        const toolbarProps = this.props.viewOption === "standard" ? this.getDefaultToolbar() : this.props.customViews;

        return toolbarProps.map(customView => ({ ...customView, onClickToolbarButton: this.onRangeChange }));
    }

    private allDayAccessor = (event: Container.ViewOptions) => event.allDay;

    private eventColor = (events: CalendarEvent) => ({ style: { backgroundColor: events.color } });

    private onRangeChange = (date: object) => {
        if (this.props.onRangeChangeAction && date) {
            this.props.onRangeChangeAction(date);
        }
    };

    private onEventDrop = (eventInfo: Container.EventInfo) => {
        if (
            eventInfo.start.getDate() !== eventInfo.event.start.getDate() &&
            this.props.editable === "default" &&
            this.props.onEventDropAction
        ) {
            this.props.onEventDropAction(eventInfo);
        }
    };

    private onEventResize = (_resizeType: string, eventInfo: Container.EventInfo) => {
        if (
            (eventInfo.start.getDate() !== eventInfo.event.start.getDate() ||
                eventInfo.end.getDate() !== eventInfo.event.end.getDate()) &&
            eventInfo.start.getDate() < eventInfo.end.getDate() &&
            this.props.editable &&
            this.props.onEventResizeAction
        ) {
            this.props.onEventResizeAction(eventInfo);
        }
    };

    private onSelectEvent = (eventInfo: object) => {
        if (this.props.onSelectEventAction) {
            this.props.onSelectEventAction(eventInfo);
        }
    };

    private onSelectSlot = (slotInfo: object) => {
        if (this.props.onSelectSlotAction) {
            this.props.onSelectSlotAction(slotInfo);
        }
    };
}

export { Calendar, Calendar as MyCalendar };
