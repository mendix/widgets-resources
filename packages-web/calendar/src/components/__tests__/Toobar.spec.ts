import { shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";

import CustomToolbar from "../Toolbar";
import { CalendarProps } from "../Calendar";

describe("Toolbar", () => {
    const renderToolbar = (props: CalendarProps): ShallowWrapper<any, any> =>
        shallow(createElement(CustomToolbar as any, props));
    const toolbarProps: CalendarProps = {
        customViews: [],
        defaultView: "month",
        enableCreate: true,
        events: [],
        formats: {},
        height: 580,
        heightUnit: "pixels",
        messages: [],
        popup: true,
        editable: "default",
        style: {},
        viewOption: "standard",
        width: 100,
        widthUnit: "percentage"
    };

    it("renders the standard toolbar structure correctly", () => {
        const toolbar = renderToolbar(toolbarProps);

        expect(toolbar).toMatchSnapshot();
    });
});
