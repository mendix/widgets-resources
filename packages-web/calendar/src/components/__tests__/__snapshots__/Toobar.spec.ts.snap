// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar renders the standard toolbar structure correctly 1`] = `
<div
  className="calendar-toolbar"
>
  <div
    className="align-left btn-group"
  />
  <div
    className="align-center btn-group"
  />
  <div
    className="align-right btn-group"
  />
</div>
`;
